### PR TITLE
(PDK-1612) Add PDK::Context and context detection

### PR DESCRIPTION
**Pull request by glennsarti**
Previously the PDK assumed it was always within a Puppet Module, but the PDK can
be used in far more different contexts.  This commit implements the PDK Context
RFC [1], in preparation for other parts of the PDK to consume it, for example
PDK validators:

* Adds the base Context and three concrete contexts: None, Module and
  ControlRepo
* Adds a memoized context object to PDK (PDK.context), for consumption by other
  parts of the PDK
* Updates the module pdk compatibility methods to accept a module path instead
  of assume the PDK is always in a module. This is needed for the Module context
  to check if an arbitrary module is PDK compatible
* Adds two test fixtures (a Puppet module and a control repo) for use by rspec
  for contex detection
* Adds tests for context detection via PDK::Context.create

[1] https://github.com/puppetlabs/pdk-planning/blob/master/RFCs/0007-add-pdk-context.md

### CODE REVIEW

File: lib/pdk/context.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      return PDK::Context::None.new(context_path) unless PDK::Util::Filesystem.directory?(context_path)
+
+      previous = nil
+      current = PDK::Util::Filesystem.expand_path(context_path)
+      until !PDK::Util::Filesystem.directory?(current) || current == previous
+        # Control Repo detection
+        return PDK::Context::ControlRepo.new(current, context_path) if PDK::ControlRepo.control_repo_root?(current)
+
+        # Puppet Module detection
+        # Note - The metadata.json file check should really be in PDK::Util.in_module_root?
```
Comment (rodjek) [inline lib/pdk/context.rb]: This was done intentionally so that modules without a `metadata.json` could be converted.

File: lib/pdk/util.rb [hunk 2]
```diff
@@ -248,16 +248,18 @@ def targets_relative_to_pwd(targets)
     module_function :targets_relative_to_pwd
 
     # TO-DO: Refactor replacement of lib/pdk/module/build.rb:metadata to use this function instead
-    def module_metadata
+    # @param module_path [String] The path to the root of the module. Default is determine the module root automatically
+    def module_metadata(module_path = nil)
       require 'pdk/module/metadata'
-
-      PDK::Module::Metadata.from_file(File.join(module_root, 'metadata.json')).data
+      module_path = module_root if module_path.nil?
```
Comment (rodjek) [inline lib/pdk/util.rb]:
```suggestion
      module_path ||= module_root
```

Or just use `module_path || module_root` in the `File.join` below

File: lib/pdk/context.rb (hunk 1 above)
Comment (glennsarti) [reply lib/pdk/context.rb]: Ok.  Will remove the comment

File: lib/pdk/util.rb (hunk 2 above)
Comment (glennsarti) [reply lib/pdk/util.rb]: Will change.

File: lib/pdk/context.rb (hunk 1 above)
Comment (glennsarti) [reply lib/pdk/context.rb]: Hrmm... so `.in_module_root?` only checks for common directories. So adding a check for metadata.json wouldn't affect that scenario.

In fact in a empty directory with only `metadata.json`:
`PDK::Util.module_root` will return the module path but
`PDK::Util.in_module_root?(..)` for that same path will return false which is "odd"

Comment (glennsarti) [reply lib/pdk/context.rb]: Removed the comment.  Can debate this elsewhere.

File: lib/pdk/util.rb (hunk 2 above)
Comment (glennsarti) [reply lib/pdk/util.rb]: Fixed

**Pull request merged**